### PR TITLE
feat(forge): use signature instead of function name

### LINKED ISSUE #977: Support overloaded functions in gas report
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.1.0 (db909f2 2022-03-18T00:07:32.053844+00:00)

### What command(s) is the bug in?

forge test --gas-report

### Operating System

macOS (amd)

### Describe the bug

Not sure if this is a bug or feature request. I have an overloaded function called getHp that accepts either one or three parameters. I see that the gas results are combined into one function.

Heres an example gas report with getHp as an overloaded function:
image

and heres me splitting getHp into two separate functions (getHp and getHp1), this is more how I would expect forge to show the gas cost for an overloaded function:
image

### PR DESCRIPTION
**Pull request by dsam82**
## Motivation

Resolves #977

## Solution

Use function signature instead of names to record calls.

> NOTE: This also changes traces, is this ok? or is there a better solution for this.

## Example

Screenshot 2022-07-17 at 18 12 40

### CODE REVIEW

**Comment by onbjerg** [comment]
I think this clutters traces a lot, perhaps passing around `Function` or a slimmed-down version of that instead would make sense? Then for gas reports you could call `.signature()` and for traces we could keep what we have now. No strong feelings, will defer to others on this

Note that this currently introduces a bug where functions named `setUp*` are not shown in gas reports

**Comment by mds1** [comment]
Agreed, not a fan of the extra clutter in the trace and would prefer just the function name there.

For gas reports I think I'd still prefer to only show the function name if no overloads, and the full sig if there is an overload, but I feel less strongly here and always showing full sig is ok with me. @transmissions11 might have some thoughts here.

Regardless, I think we should show the full sig as `a(uint256)(uint256)`, which is how you pass a sig to cast, instead of the current colon-separated version of `a(uint256):(uint256)`

**Comment by transmissions11** [comment]
> For gas reports I think I'd still prefer to only show the function name if no overloads, and the full sig if there is an overload, but I feel less strongly here and always showing full sig is ok with me. @transmissions11 might have some thoughts here.

Overall agreed @mds1. Strongly against doing the full sig by default, but showing it in case of overload could make sense.

**Comment by dsam82** [comment]
Thanks for the feedback. Had same thoughts, so went with minimal working changes. 
Will look into this more.

> Note that this currently introduces a bug where functions named setUp* are not shown in gas reports

Oops 🙂, will change this.

**Comment by dsam82** [comment]
@onbjerg, @mds1 would love your review again.

Have updated `RawOrDecoded::Decoded` tuple to `(name, sig, inputs)` and then used that in traces and gas-reports.

For only showing full signature when overloaded, updated `ContractInfo::functions` with another nested map.

**Comment by mds1** [comment]
That looks great to me, thank you!

File: forge/src/gas_report.rs [hunk 1]
```diff
@@ -65,12 +65,14 @@ impl GasReport {
...
-                    RawOrDecodedCall::Decoded(func, _)
+                    RawOrDecodedCall::Decoded(func, sig, _)
                         if !func.starts_with("test") && func != "setUp" =>
                     {
                         let function_report = contract_report
                             .functions
                             .entry(func.clone())
+                            .or_insert_with(Default::default)
+                            .entry(sig.clone())
                             .or_insert_with(Default::default);
```
Comment (onbjerg) [inline forge/src/gas_report.rs:76]: I think same as above (my bad)

**Review by onbjerg — APPROVED** [review]
minor nits otherwise LGTM

**Review by mattsse — APPROVED** [review]
lgtm, ty

**Pull request merged**